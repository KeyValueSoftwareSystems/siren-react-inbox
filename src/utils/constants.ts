export const COLORS = {
  light: {
    primaryColor: "#F56630",
    highlightedCardColor: "#FFECE5",
    textColor: "#344054",
    neutralColor: "#FFFFFF",
    borderColor: "#D0D5DD",
    dateColor: "#34405499",
    deleteIcon: "#34405499",
    timerIcon: "#667185",
    clearAllIcon: "#667185",
    windowShadowColor: "#D0D5DD",
    iconColor: "#F7F9FC",
    notificationIcon: "#232326",
  },
  dark: {
    primaryColor: "#F56630",
    highlightedCardColor: "#2E2D30",
    textColor: "#FFFFFF",
    neutralColor: "#232326",
    borderColor: "#344054",
    dateColor: "#98A2B3",
    deleteIcon: "#D0D5DD",
    timerIcon: "#98A2B3",
    clearAllIcon: "#D0D5DD",
    windowShadowColor: "#232326",
    iconColor: "#38383D",
    notificationIcon: "#FFFFFF",
  },
};

export enum ThemeMode {
  DARK = "dark",
  LIGHT = "light",
}

export enum LogLevel {
  INFO,
  ERROR,
}

export enum BadgeType {
  NONE = "none",
  DOT = "dot",
  DEFAULT = "default",
}

export const levelLogFns = {
  [LogLevel.INFO]: console.log,
  [LogLevel.ERROR]: console.error,
};

export const defaultBadgeStyle = {
  size: 15,
  color: "#FF0000",
  textColor: "#FFFFFF",
  textSize: 10,
  linHeight: "14px",
  top: "inherit",
  left: "inherit",
};

export enum eventTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  UPDATE_NOTIFICATIONS_COUNT = "UPDATE_NOTIFICATIONS_COUNT",
  RESET_NOTIFICATIONS = "RESET_NOTIFICATIONS",
  RESET_NOTIFICATIONS_COUNT = "RESET_NOTIFICATIONS_COUNT",
}

export enum events {
  NOTIFICATION_LIST_EVENT = "NOTIFICATION_LIST_EVENT",
  NOTIFICATION_COUNT_EVENT = "NOTIFICATION_COUNT_EVENT",
}

export enum updateNotificationsTypes {
  MARK_ITEM_AS_VIEWED = "MARK_ITEM_AS_VIEWED",
  MARK_ITEM_AS_READ = "MARK_ITEM_AS_READ",
  MARK_ALL_AS_READ = "MARK_ALL_AS_READ",
  DELETE_ITEM = "DELETE_ITEM",
  DELETE_ALL_ITEM = "DELETE_ALL_ITEM",
}

export enum sirenReducerTypes {
  SET_NOTIFICATIONS = "SET_NOTIFICATIONS",
  NEW_NOTIFICATIONS = "NEW_NOTIFICATIONS",
  SET_UN_VIEWED_NOTIFICATION_COUNT = "SET_UN_VIEWED_NOTIFICATION_COUNT",
  SET_SIREN_CORE = "SET_SIREN_CORE",
}

export const LIST_EMPTY_TEXT = "No new notifications";
export const LIST_EMPTY_SUB_TEXT = "Check back later for updates and alerts.";
export const ERROR_TEXT = "Oops! Something went wrong";
export const ERROR_SUB_TEXT =
  "Could not load the notifications. Please refresh the page.";
export const DEFAULT_WINDOW_TITLE = "Notifications";
export const RETRY_BUTTON_LABEL = "Retry";
export const CLEAR_ALL_LABEL = "Clear All";
export const IN_APP_RECIPIENT_UNAUTHENTICATED = 'IN_APP_RECIPIENT_UNAUTHENTICATED';
export const TOKEN_VERIFICATION_PENDING ='TOKEN_VERIFICATION_PENDING';
export const MAXIMUM_RETRY_COUNT = 3;

export const errorMap = {
  SIREN_OBJECT_NOT_FOUND: {
    Type: "ERROR",
    Code: "SIREN_OBJECT_NOT_FOUND",
    Message: "Siren Object Not found",
  },
  MISSING_PARAMETER: {
    Type: "ERROR",
    Code: "MISSING_PARAMETER",
    Message: "Missing Parameter",
  },
};

export enum VerificationStatus {
  PENDING = "PENDING",
  SUCCESS = "SUCCESS",
  FAILED = "FAILED"
}