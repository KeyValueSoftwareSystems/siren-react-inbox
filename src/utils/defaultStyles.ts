
const defaultStyles = {
  notificationIcon: {
    size: 30,
  },
  window: {
    width: 500,
    borderRadius: "10px",
    borderWidth: "0px",
    borderStyle: "solid"
  },
  windowHeader: {
    height: 64,
    titleFontWeight: "600",
    titleSize: 20,
    closeIconSize: 20,
    titlePadding: 0,
  },
  windowContainer: {
    padding: 0,
    height: 700,
  },
  notificationCard: {
    padding: 10,
    borderWidth: 0.5,
    avatarSize: 40,
    titleFontWeight: "600",
    titleSize: 16,
    descriptionSize: 14,
    dateSize: 12,
  },
  loadMoreButton: {
    fontSize: 14,
    fontWeight: '600'
  },
  badgeStyle: {
    size: 15,
    textSize: 10,
  },
  deleteIcon:{
    size: 16
  },
  timerIcon:{
    size: 12
  },
  clearAllIcon:{
    size: 18
  }
}

export default defaultStyles;
