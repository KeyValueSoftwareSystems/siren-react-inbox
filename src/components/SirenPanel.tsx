import React, { type FC, useCallback, useEffect, useMemo, useState } from "react";

import type {
  ActionResponse,
  MarkAsViewedResponse,
  NotificationDataType,
  NotificationsApiResponse,
  SirenErrorType,
} from "@sirenapp/js-sdk/dist/esm/types";

import "../styles/sirenPanel.css";
import NotificationCard from "./Card";
import EmptyList from "./EmptyList";
import ErrorWindow from "./ErrorWindow";
import Header from "./Header";
import AnimatedLoader from "./Loader";
import ShowMoreButton from "./ShowMore";
import { useSirenContext } from "./SirenProvider";
import Tab from "./Tab";
import type { EventListenerDataType, SirenPanelProps } from "../types";
import {
  filterDataProperty,
  generateFilterParams,
  getModalContentHeightInFullScreen,
  isEmptyArray,
  isValidResponse,
  mergeArrays,
  mergeStyles,
  updateNotifications,
} from "../utils/commonUtils";
import { DEFAULT_WINDOW_TITLE, ERROR_TEXT, errorMap, events, EventType, eventTypes, VerificationStatus } from "../utils/constants";
import useSiren from "../utils/sirenHook";

/**
 * SirenPanel component renders a notification panel with a header, notification cards, and optional custom footer.
 *
 * @component
 * @example
 * <SirenPanel
 *   styles={customStyles}
 *   title="Notifications"
 *   hideHeader={false}
 *   cardProps={{ hideAvatar: false, showMedia: true }}
 *   renderListEmpty={() => <div>No notifications</div>}
 *   customFooter={<FooterComponent />}
 *   customHeader={<CustomHeader />}
 *   customCard={(dataItem) => <CustomNotificationCard data={dataItem} />}
 *   onCardClick={(notification) => console.log('Notification clicked', notification)}
 * />
 *
 * @param {SirenPanelProps} props - The properties passed to the SirenWindow component.
 * @param {Object} props.styles - Custom styles applied to the notification panel and its elements.
 * @param {string} props.loadMoreLabel - Label for load more button  
 * @param {boolean} [props.hideBadge] - Flag indicating if the badge should be hidden
 * @param {Object} props.headerProps - Object containing props related to the inbox header.
 * @param {Object} props.cardProps - Optional properties to customize the appearance of notification cards.
 * @param {boolean} props.darkMode - Flag indicating if the component is in dark mode.
 * @param {number} props.noOfNotificationsPerFetch - The number of notifications to fetch per request.
 * @param {boolean} props.fullScreen - Flag indicating if the component is in full screen mode.
 * @param {ReactNode} props.listEmptyComponent - Custom component to be rendered when the notification list is empty.
 * @param {ReactNode} props.customFooter - Custom footer component to be rendered below the notification list.
 * @param {ReactNode} pros.customLoader - Custom Loader component to be rendered while fetching notification list for the first time
 * @param {ReactNode} pros.loadMoreComponent -Custom load more component to be rendered
 * @param {ReactNode} props.customErrorWindow -Custom error window component to be rendered when there is an error
 * @param {Function} props.customCard - Function to render custom notification cards.
 * @param {Function} props.onCardClick - Callback function executed when a notification card is clicked.
 * @param {DimensionValue} props.modalWidth - The width of the notification panel.
 * @param {Function} props.onError - Callback function executed when an error occurs.
 * @returns {ReactElement} The rendered SirenInbox component.
 */


const SirenPanel: FC<SirenPanelProps> = ({
  styles,
  loadMoreLabel,
  hideBadge,
  headerProps,
  cardProps,
  darkMode,
  noOfNotificationsPerFetch,
  fullScreen,
  listEmptyComponent,
  customFooter,
  customLoader,
  loadMoreComponent,
  customErrorWindow,
  customCard,
  onCardClick,
  onError,
  modalWidth,
  hideTab = false,
  tabProps = {
    tabs: [
      { key: 'All', title: 'All' },
      { key: 'Unread', title: 'Unread' }
    ],
    activeTab: 0
  }
}) => {
  const {
    markAllAsViewed,
    deleteByDate,
    deleteById,
  } = useSiren();
  const { siren, verificationStatus, id } = useSirenContext();
  const {hideHeader = false, hideClearAll = false, customHeader, title = DEFAULT_WINDOW_TITLE} = headerProps ?? {};
  const [notifications, setNotifications] = useState<NotificationDataType[]>(
    []
  );
  const [activeTabIndex, setActiveTabIndex] = useState(0);
  const [filterType, setFilterType] = useState<string>(tabProps.tabs[tabProps.activeTab].key);
  const [error, setError] = useState<string>("");
  const [isLoading, setIsLoading] = useState(true);
  const [endReached, setEndReached] = useState(false);
  const [eventListenerData, setEventListenerData] =
    useState<EventListenerDataType | null>(null);

  const notificationSubscriber = async (type: string, dataString: string) => {
    const data = await JSON.parse(dataString);

    setEventListenerData(data);
  };  

  useEffect(() => {
    PubSub.subscribe(`${events.NOTIFICATION_LIST_EVENT}${id}`, notificationSubscriber);

    return () => {
      cleanUp();
      setNotifications([]);
      handleMarkNotificationsAsViewed(new Date().toISOString());
    };
  }, [siren]);

  useEffect(() => {
    return(() => {
      !hideBadge && restartNotificationCountFetch();
    })
  }, [hideBadge])

  useEffect(() => {
    if (eventListenerData) {
      const updatedNotifications: NotificationDataType[] = updateNotifications(
        eventListenerData,
        notifications,
      );

      if(!isEmptyArray(eventListenerData?.newNotifications)) handleMarkNotificationsAsViewed(updatedNotifications[0]?.createdAt);
      setNotifications(updatedNotifications);
      setEventListenerData(null);
    }
  }, [eventListenerData]);

  const handleVerificationFailure = () => {
    setIsLoading(false);
    onError && onError(errorMap?.INVALID_CREDENTIALS);
    setError(ERROR_TEXT);
  };

  useEffect(() => {
    if (siren && verificationStatus === VerificationStatus.SUCCESS) {
      !hideBadge && siren.stopRealTimeFetch(EventType.UNVIEWED_COUNT);
      fetchNotifications(true);
    }
    else if(verificationStatus === VerificationStatus.FAILED) {
      handleVerificationFailure();
    }
  }, [siren, verificationStatus, hideBadge, activeTabIndex]);

  const restartNotificationCountFetch = () => {
    try {
      siren?.startRealTimeFetch({eventType: EventType.UNVIEWED_COUNT});
    } catch (er) {
      //  handle error if needed
    }
  };

  const cleanUp = () => {
    siren?.stopRealTimeFetch(EventType.NOTIFICATION);
  };

  const triggerOnError = useCallback(
    (
      response: NotificationsApiResponse | ActionResponse | MarkAsViewedResponse
    ) => {
      if (response?.error && onError) onError(response.error);
    },
    [onError]
  );

  const resetValues = () => {
    setNotifications([]);
    setEndReached(false);
    setError("");
  };

  const handleClearAllNotification = async (): Promise<void> => {   
    try {
      if (!isEmptyArray(notifications)) {
        const response = await deleteByDate(
          notifications[0].createdAt
        );

        response && triggerOnError(response);

        if (response && isValidResponse(response)) {
          resetValues();
          setEndReached(true);
          setIsLoading(false);
        }
      }
    } catch (er) {
      //  handle error if needed
    }
  };

  const updateNotificationList = (
    newNotifications: NotificationDataType[],
    isRefresh: boolean
  ) => {
    const updatedNotifications = !isRefresh
      ? mergeArrays(notifications, newNotifications)
      : newNotifications;

    setNotifications(updatedNotifications);

    if (isRefresh)
      handleMarkNotificationsAsViewed(updatedNotifications[0].createdAt);
  };

  const fetchNotifications = async (isRefresh = false) => {
    setError("");
    setIsLoading(true);

    if (!siren) return [];

    try {
      const response = await siren.fetchAllNotifications(
        generateFilterParams(
          isRefresh ? [] : notifications,
          false,
          noOfNotificationsPerFetch,
          filterType
        )
      );

      if (response && isValidResponse(response)) {
        let data: NotificationDataType[] | null = null;

        if (!isEmptyArray(response.data ?? [])) {
          data = filterDataProperty(response);
          if (!data) return [];
          if (response?.meta) {
            const isLastPage = response?.meta?.last === "true";

            if (isLastPage) setEndReached(true);
            else setEndReached(false);
          }
          updateNotificationList(data, isRefresh);
        }
        if (!data) setEndReached(true);
        resetRealTimeFetch(isRefresh, data);
      } else {
        setEndReached(true);
        setError(ERROR_TEXT);
      }
    }  catch (error) {
      setIsLoading(false);
      if (typeof error === 'object' && error !== null && 'Message' in error) 
        setError((error as SirenErrorType).Message);
      else 
        setError(ERROR_TEXT);     
    }

    setIsLoading(false);
  };

  const resetRealTimeFetch = (
    refresh: boolean,
    newList: NotificationDataType[] | null
  ) => {
    if (!refresh) return;

    try {
      siren?.startRealTimeFetch(
        {eventType: EventType.NOTIFICATION, params:   generateFilterParams(newList ?? [], true, noOfNotificationsPerFetch, filterType)}   
      );
    } catch (er) {
      //  handle error if needed
    }
  };

  const deleteNotificationById = useCallback(
    async (id: string, shouldUpdateList: boolean) => {
      let isSuccess = false;

      try {
        const response = await deleteById(id, shouldUpdateList);

        if (response?.data) isSuccess = true;
        response && triggerOnError(response);
      } catch (er) {
        isSuccess = false;
        //  handle error if needed
      }

      return isSuccess;
    },
    [deleteById, triggerOnError]
  );

  const onEndReached = (): void => {
    if (!isLoading && !endReached && notifications?.length > 0)
      fetchNotifications(false);
  };

  const handleMarkNotificationsAsViewed = async (
    createdAt: string
  ): Promise<void> => {
    try {
      const payload = {
        unviewedCount: 0,
        action: eventTypes.UPDATE_NOTIFICATIONS_COUNT,
      };

      PubSub.publish(`${events.NOTIFICATION_COUNT_EVENT}${id}`, JSON.stringify(payload));
      if (createdAt) {
        const response = await markAllAsViewed(createdAt);

        response && triggerOnError(response);
      }
    } catch (er) {
      //  handle error if needed
    }
  };

  const handleLoadMore = (event: React.MouseEvent) => {
    event.preventDefault();
    onEndReached();
  };


  const handleTabChange = (index: number) => {
    resetValues();
    cleanUp();
    setActiveTabIndex(index);
    setFilterType(tabProps.tabs[index].key);
  };

  const renderFooter = () => {
    if (isEmptyArray(notifications) && isLoading) return <div />;

    return (
      <div className="siren-sdk-panel-footer" data-testid="custom-footer">
        {customFooter}
      </div>
    );
  };

  const renderedListItems = useMemo(() => {
    return notifications.map((item) => (
      <NotificationCard
        notification={item}
        cardProps={cardProps}
        onCardClick={onCardClick}
        deleteNotificationById={deleteNotificationById}
        styles={styles}
        key={item.id}
        darkMode={darkMode}
        data-testid="notification-card"
      />
    ));
  }, [notifications, cardProps, onCardClick, deleteNotificationById, styles, darkMode]);
  

  const renderList = () => {
    if (isLoading && isEmptyArray(notifications)) {
      const hideAvatar = cardProps?.hideAvatar || false;
      
      return (
        <div className="siren-sdk-list-loader-container">
          {customLoader || 
          <>
            <AnimatedLoader styles={styles} hideAvatar={hideAvatar}/>
            <AnimatedLoader styles={styles} hideAvatar={hideAvatar}/>
            <AnimatedLoader styles={styles} hideAvatar={hideAvatar}/>
            <AnimatedLoader styles={styles} hideAvatar={hideAvatar}/>
            <AnimatedLoader styles={styles} hideAvatar={hideAvatar}/>
          </>}
        </div>
      );
    }

    if (error && !isLoading)
      return (
        <div aria-label="siren-error-state">
          { customErrorWindow || (
            <ErrorWindow styles={styles} error={error} darkMode={darkMode} />
          )}
        </div>
      );

    if (isEmptyArray(notifications) && !error && !isLoading)
      return (
        <div aria-label="siren-empty-state">
          {listEmptyComponent || (
            <EmptyList
              data-testid="empty-list"
              styles={styles}
              darkMode={darkMode}
            />)
          }
        </div>     
      );

    if (customCard) 
      return notifications.map((item) => customCard(item)); 
     
    return renderedListItems;
  };

  const renderListBottomComponent = useMemo(() => {
    if (isEmptyArray(notifications)) return null;
    if (isLoading && !endReached)
      return (
        <div className="siren-sdk-panel-infinite-loader-container">
          <div
            className="siren-sdk-panel-infinite-loader"
            style={styles.infiniteLoader}
          />
        </div>
      );
    if (!isLoading && !endReached)
      return (
        <ShowMoreButton
          styles={styles}
          customComponent={loadMoreComponent}
          onClick={handleLoadMore}
          loadMoreLabel={loadMoreLabel}
        />
      );

    return null;
  }, [notifications, isLoading, endReached, styles, loadMoreComponent, handleLoadMore, loadMoreLabel]);


  const containerClassNames = useMemo(() => {
    const hasNotifications = notifications?.length;
    const hasError = error;
    const isEmptyOrErrorState = !hasNotifications || hasError && !isLoading;
  
    const classes = [];

    if (isEmptyOrErrorState) classes.push("siren-sdk-content-container");
    if (customFooter) classes.push("siren-sdk-panel-no-border");
  
    return classes.join(' ');
  }, [notifications, error, isLoading, customFooter]);
  
  
  const panelStyle = useMemo(() => {
    let baseStyles = { ...styles.container };
  
    if (!fullScreen) 
      baseStyles = mergeStyles(
        baseStyles,
        styles.windowTopBorder,
        { width: `${modalWidth}px` },
        styles.windowBottomBorder
      );
    
    
    return baseStyles;
  }, [fullScreen, styles, modalWidth]);
  

  const contentContainerHeightInFullScreen = useMemo(() => {
    return getModalContentHeightInFullScreen(styles?.headerContainer?.height);
  }, [styles?.headerContainer?.height]);

  const defaultTabProps = {
    tabs: tabProps.tabs,
    activeTab: tabProps.activeTab,
    onTabChange: handleTabChange, 
    styles: styles,
  };

  return (
    <div
      className={
        !fullScreen ? "siren-sdk-panel-modal" : "siren-sdk-panel-container"
      }
      style={panelStyle}
      data-testid="siren-panel"
    >
      <div className={fullScreen ? 'siren-sdk-panel-modal-fullscreen' : ''}>
        {!hideHeader &&
          (customHeader || (
            <Header
              title={title}
              styles={styles}
              hideClearAll={hideClearAll}
              fullScreen={fullScreen}
              enableClearAll={!isEmptyArray(notifications)}
              handleClearAllNotification={handleClearAllNotification}
            />
          ))}
        <div
          style={{
            ...(!fullScreen && styles.windowBottomBorder),
            ...(fullScreen && {height: contentContainerHeightInFullScreen}),
            ...styles.contentContainer,
          }}
        >
          {!hideTab && (<Tab {...defaultTabProps} />)} 
          <div
            id="contentContainer"
            style={{
              ...(fullScreen ? { height: contentContainerHeightInFullScreen } : {
                ...styles.windowBottomBorder,
                ...styles.body
              }),
            }}
            className={`siren-sdk-panel-content-container ${containerClassNames}`}
            aria-label="siren-notification-list"
          >
            {renderList()}
            {renderListBottomComponent}
          </div>
          {renderFooter()}
        </div>
      </div>
    </div>
  );
};

export default SirenPanel;
