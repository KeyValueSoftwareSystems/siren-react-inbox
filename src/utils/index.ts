import * as CommonUtils from './commonUtils';
import * as Constants from './constants';
import DefaultStyle from './defaultStyles'
import DefaultTheme from './defaultTheme';
import useSiren from './sirenHook';

export {
  Constants,
  DefaultTheme,
  useSiren,
  CommonUtils,
  DefaultStyle
};