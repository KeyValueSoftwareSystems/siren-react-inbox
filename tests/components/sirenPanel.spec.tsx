import React from "react";

import { render, waitFor } from "@testing-library/react";

import SirenPanel from "../../src/components/SirenPanel"; // Adjust the import path accordingly
import { applyTheme } from "../../src/utils/commonUtils";

// Mock the CSS files to avoid Jest error
jest.mock("../../src/styles/sirenNotificationIcon.css", () => ({}));
jest.mock("../../src/styles/header.css", () => ({}));
jest.mock("../../src/styles/card.css", () => ({}));
jest.mock("../../src/styles/loader.css", () => ({}));
jest.mock("../../src/styles/emptyList.css", () => ({}));
jest.mock("../../src/styles/errorWindow.css", () => ({}));
jest.mock("../../src/styles/sirenPanel.css", () => ({}));
jest.mock("../../src/styles/showMore.css", () => ({}));

const mockErrorFn = jest.fn();
const style = applyTheme();

const props = {
  styles: style,
  title: "Notifications",
  inboxHeaderProps: { hideHeader: false},
  cardProps: { hideAvatar: false, showMedia: true },
  customFooter: undefined,
  customHeader: undefined,
  onNotificationCardClick: undefined,
  customNotificationCard: undefined,
  onError: mockErrorFn,
  listEmptyComponent: undefined,
  noOfNotificationsPerFetch: 10,
  fullScreen: false,
  setModalVisible: jest.fn(),
  darkMode: false,
  modalWidth: 500,
};

test("matches snapshot", () => {
  const { asFragment } = render(<SirenPanel {...props} />);

  expect(asFragment()).toMatchSnapshot();
});

it("renders title when provided", () => {
  const { getByText } = render(<SirenPanel {...props} inboxHeaderProps={{title: "Notifications"}} />);
  const title = getByText("Notifications");

  expect(title).toBeTruthy();
});

it("hides header when hideHeader prop is true", () => {
  const inboxHeaderProps = {hideHeader: true}
  const { queryByText } = render(<SirenPanel {...props} inboxHeaderProps={inboxHeaderProps} />);
  const header = queryByText("Notifications");

  expect(header).toBeFalsy();
});

it("renders custom header when provided", () => {
  const customHeader = <div data-testid="custom-header">Custom Header</div>;
  const inboxHeaderProps = {customHeader}
  const { getByTestId } = render(
    <SirenPanel {...props} inboxHeaderProps={inboxHeaderProps} />
  );
  const header = getByTestId("custom-header");

  expect(header).toBeTruthy();
});

it("calls onError callback if an error occurs", async () => {
  // Mocking an error response from the server
  jest.spyOn(global, "fetch").mockRejectedValueOnce({
    json: async () => ({
      Code: "OUTSIDE_SIREN_CONTEXT",
      Message: "Trying to invoke function outside the siren context",
      Type: "ERROR",
    }),
  });
  render(<SirenPanel {...props} />);
  await waitFor(() => {
    expect(mockErrorFn).toHaveBeenCalledWith({
      Code: "OUTSIDE_SIREN_CONTEXT",
      Message: "Trying to invoke function outside the siren context",
      Type: "ERROR",
    });
  });
});
