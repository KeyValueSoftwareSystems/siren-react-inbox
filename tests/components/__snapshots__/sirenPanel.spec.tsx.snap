// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="siren-sdk-panel-modal"
    data-testid="siren-panel"
    style="max-width: 100; border-top-left-radius: 10px; border-top-right-radius: 10px; width: 500px; border-bottom-left-radius: 10px; border-bottom-right-radius: 10px;"
  >
    <div
      class=""
    >
      <div
        class="siren-sdk-header-container"
        data-testid="header"
        style="border-top-left-radius: 10px; border-top-right-radius: 10px; background-color: rgb(35, 35, 38); border-bottom: 0.5px solid; border-color: #344054; height: 50px;"
      >
        <p
          class="siren-sdk-text-break"
          style="color: rgb(255, 255, 255); font-size: 18px; font-weight: 600; padding-left: 0px;"
        >
          Notifications
        </p>
        <div
          aria-disabled="true"
          aria-label="siren-header-clear-all"
          class="siren-sdk-header-right-container"
          data-testid="clear-all"
          role="button"
          style="opacity: 0.5; cursor: default;"
        >
          <svg
            fill="none"
            height="18"
            viewBox="0 0 24 24"
            width="18"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5 12C5 12.5523 5.44772 13 6 13H18C18.5523 13 19 12.5523 19 12C19 11.4477 18.5523 11 18 11H6C5.44772 11 5 11.4477 5 12ZM3 16C3 16.5523 3.44772 17 4 17H16C16.5523 17 17 16.5523 17 16C17 15.4477 16.5523 15 16 15H4C3.44772 15 3 15.4477 3 16ZM8 7C7.44772 7 7 7.44772 7 8C7 8.55228 7.44772 9 8 9H20C20.5523 9 21 8.55228 21 8C21 7.44772 20.5523 7 20 7H8Z"
              fill="#D0D5DD"
            />
          </svg>
          <p
            class="siren-sdk-header-clear-all-text"
            style="color: rgb(255, 255, 255);"
          >
            Clear All
          </p>
        </div>
      </div>
      <div
        style="border-bottom-left-radius: 10px; border-bottom-right-radius: 10px; background-color: rgb(35, 35, 38); padding: 0px; border-radius: 0 0 10px 10px;"
      >
        <div
          class="siren-sdk-tab-container"
        >
          <div
            class="siren-sdk-tab-header-container"
            style="height: 50px; background-color: rgb(35, 35, 38);"
          >
            <div
              class="siren-sdk-tab-header active"
              role="button"
              style="margin-right: 22px; background-color: rgb(35, 35, 38); color: rgb(250, 152, 116); font-size: 14px; font-weight: 600;"
            >
              All
            </div>
            <div
              class="siren-sdk-tab-header "
              role="button"
              style="margin-right: 0px; background-color: rgb(35, 35, 38); color: rgb(255, 255, 255); font-size: 14px; font-weight: 600;"
            >
              Unread
            </div>
            <div
              class="siren-sdk-active-tab-indicator"
              style="background-color: rgb(250, 152, 116); height: 3px; left: 0px; width: 0px;"
            />
          </div>
        </div>
        <div
          aria-label="siren-notification-list"
          class="siren-sdk-panel-content-container siren-sdk-content-container"
          id="contentContainer"
          style="border-bottom-left-radius: 10px; border-bottom-right-radius: 10px; height: 700px;"
        >
          <div
            class="siren-sdk-list-loader-container"
          >
            <div
              class="siren-sdk-skeleton-container"
            >
              <div
                class="siren-sdk-skeleton-grid-with-avatar"
              >
                <div
                  class="siren-sdk-skeleton-avatar siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-head siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-subtitle siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-body siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-icon  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-footer  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-close  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
              </div>
            </div>
            <div
              class="siren-sdk-skeleton-container"
            >
              <div
                class="siren-sdk-skeleton-grid-with-avatar"
              >
                <div
                  class="siren-sdk-skeleton-avatar siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-head siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-subtitle siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-body siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-icon  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-footer  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-close  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
              </div>
            </div>
            <div
              class="siren-sdk-skeleton-container"
            >
              <div
                class="siren-sdk-skeleton-grid-with-avatar"
              >
                <div
                  class="siren-sdk-skeleton-avatar siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-head siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-subtitle siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-body siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-icon  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-footer  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-close  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
              </div>
            </div>
            <div
              class="siren-sdk-skeleton-container"
            >
              <div
                class="siren-sdk-skeleton-grid-with-avatar"
              >
                <div
                  class="siren-sdk-skeleton-avatar siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-head siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-subtitle siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-body siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-icon  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-footer  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-close  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
              </div>
            </div>
            <div
              class="siren-sdk-skeleton-container"
            >
              <div
                class="siren-sdk-skeleton-grid-with-avatar"
              >
                <div
                  class="siren-sdk-skeleton-avatar siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-head siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-subtitle siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-body siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-icon  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-footer  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
                <div
                  class="siren-sdk-skeleton-close  siren-sdk-skeleton"
                  style="border-color: #fa9874;"
                />
              </div>
            </div>
          </div>
        </div>
        <div />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
